<a href="https://colab.research.google.com/github/sebastianhariman/NewsWebScraping/blob/main/News%20Scraping%20and%20Classification%20Using%20Beautifulsoup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **News Scraping and Classification Using Beautifulsoup**





In [ ]:
!pip install beautifulsoup4

In [ ]:
# Import Libraries

import numpy as np
import pandas as pd

from bs4 import BeautifulSoup
import requests
import csv

## **Data Preparation**

### **Data Collection**

The first step of data preparation is collecting news urls from various media.

In [ ]:
Kompas = [['https://www.kompas.com/food/read/2023/10/30/100800175/-jakarta-coffee-week-2023-digelar-3-5-november-pencinta-kopi-mari-bersiap', 'Kompas', 'Hiburan'],
           ['https://nasional.kompas.com/read/2023/11/05/16001461/jusuf-kalla-minta-pemerintah-indonesia-rangkul-as-untuk-hentikan-serangan', 'Kompas', 'Politik'],
           ['https://www.kompas.com/badminton/read/2023/11/04/18300068/marcus-gideon-kalau-saya-maksa-barengan-terus-kasihan-kevin', 'Kompas', 'Olahraga'],
          ['https://nasional.kompas.com/read/2023/11/10/11020991/kpk-tetapkan-tersangka-korupsi-pengadaan-apd-covid-19-di-kemenkesn', 'Kompas', 'Politik'],
          ['https://www.kompas.com/food/read/2023/11/10/140800175/kisah-kopi-jos-pertama-di-yogyakarta-berdiri-sejak-1960-an', 'Kompas', 'Hiburan'],
          ['https://www.kompas.com/badminton/read/2023/11/09/11470238/hasil-korea-masters-2023-adnan-nita-dan-shesar-tersingkir', 'Kompas', 'Olahraga'],
          ['https://www.kompas.com/hype/read/2023/11/10/210111966/god-bless-raih-rekor-muri-band-rock-tertua-yang-masih-berkarya', 'Kompas', 'Hiburan'],
          ['https://bola.kompas.com/read/2023/11/10/21132308/hasil-indonesia-vs-ekuador-garuda-muda-raih-hasil-imbang', 'Kompas', 'Olahraga']]

CNN = [['https://www.cnnindonesia.com/hiburan/20231103204814-220-1019857/sinopsis-songbird-bioskop-trans-tv-4-november-2023', 'CNN Indonesia', 'Hiburan'],
           ['https://www.cnnindonesia.com/nasional/20231104204544-32-1020063/hasto-mas-gibran-tidak-lagi-anggota-pdip-karena-sudah-pamit', 'CNN Indonesia', 'Politik'],
           ['https://www.cnnindonesia.com/olahraga/20231104230035-170-1020074/hasil-hylo-open-rehan-lisa-menang-comeback-lolos-ke-final', 'CNN Indonesia', 'Olahraga'],
       ['https://www.cnnindonesia.com/olahraga/20231110151354-156-1022598/hasil-practice-motogp-malaysia-alex-marquez-tercepat-martin-kedua', 'CNN Indonesia', 'Olahraga'],
       ['https://www.cnnindonesia.com/hiburan/20231110121645-227-1022452/nct-127-akan-konser-neo-city-the-unity-di-jakarta-januari-2024', 'CNN Indonesia', 'Hiburan'],
       ['https://www.cnnindonesia.com/nasional/20231109201234-32-1022242/dpr-ingin-jokowi-lantik-panglima-tni-baru-sebelum-26-november', 'CNN Indonesia', 'Politik'],
       ['https://www.cnnindonesia.com/hiburan/20231111013712-227-1022836/daftar-lengkap-nominasi-grammy-awards-2024', 'CNN Indonesia', 'Hiburan'],
       ['https://www.cnnindonesia.com/nasional/20231110152656-32-1022605/dpr-akan-dalami-isu-pemilu-dan-papua-ke-calon-panglima-tni', 'CNN Indonesia', 'Politik']]

Detik = [['https://sport.detik.com/sepakbola/liga-jerman/d-7019587/harry-kane-sudah-15-gol-torehkan-sejarah-di-bundesliga?mtype=mpc.w.B-boxccxmpcxmp-modelB', 'Detik', 'Olahraga'],
           ['https://news.detik.com/pemilu/d-7020967/gerindra-hormati-pdip-soal-menteri-pak-jokowi-tak-berubah-sama-sekali', 'Detik', 'Politik'],
           ['https://hot.detik.com/kdrama/d-7023685/5-hal-yang-bikin-penasaran-dari-sweet-home-2', 'Detik', 'Hiburan'],
         ['https://sport.detik.com/basket/d-7017340/hasil-nba-wembanyama-gemilang-spurs-kembali-tekuk-suns', 'Detik', 'Olahraga'],
         ['https://hot.detik.com/book/d-7029566/9-tahun-tak-terbit-manga-doraemon-plus-rilis-volume-terbaru', 'Detik', 'Hiburan'],
         ['https://news.detik.com/berita/d-7026587/jokowi-akan-hadiri-ktt-oki-di-arab-saudi-bahas-gaza', 'Detik', 'Politik'],
         ['https://sport.detik.com/sepakbola/liga-inggris/d-7030650/bukan-enzo-bukan-sterling-ini-pemain-chelsea-paling-berbahaya', 'Detik', 'Olahraga'],
         ['https://news.detik.com/berita/d-7030433/kpk-geledah-rumah-ketua-komisi-iv-dpr-sudin', 'Detik', 'Politik']]

Liputan6 = [['https://www.liputan6.com/lifestyle/read/5433966/taman-hiburan-indoor-terbesar-di-dunia-dibuka-di-china-punya-akuarium-raksasa-berisi-paus-orca?page=4', 'Liputan6', 'Hiburan'],
           ['https://www.liputan6.com/bola/read/5445235/bibit-bibit-muda-basket-indonesia-diharapkan-lahir-dari-dash-basketball-tournament?page=3', 'Liputan6', 'Olahraga'],
           ['https://www.liputan6.com/news/read/5444186/unjuk-rasa-aliansi-mahasiswa-tolak-hadirnya-politik-dinasti?page=2', 'Liputan6', 'Politik'],
            ['https://www.liputan6.com/bola/read/5448360/real-madrid-sudah-temukan-target-pengganti-mbappe-ciri-cirinya-pemain-ini-pernah-sekamar-dengan-bellingham?page=4', 'Liputan6', 'Olahraga'],
            ['https://www.liputan6.com/showbiz/read/5449475/disutradarai-hanung-bramantyo-drakor-descendants-of-the-sun-bakal-diangkat-ke-layar-lebar?page=4', 'Liputan6', 'Hiburan'],
            ['https://www.liputan6.com/news/read/5449611/kata-mahfud-md-soal-tudingan-pernah-terlibat-konflik-kepentingan-saat-jadi-ketua-mk?page=3', 'Liputan6', 'Politik'],
            ['https://www.liputan6.com/showbiz/read/5449994/aruma-raih-penghargaan-ami-award-2023-melalui-lagu-muak-dinobatkan-sebagai-penyanyi-pendatang-baru-terbaik?page=4', 'Liputan6', 'Hiburan'],
            ['https://www.liputan6.com/bola/read/5448810/hasil-nba-clippers-belum-menang-juga-sejak-rekrut-james-harden?page=2', 'Liputan6', 'Olahraga']]

Dataset is collected from four news media, such as Kompas, CNN Indonesia, Detik, and Liputan6. Within each media, eight news from three categories (politics, entertainment, and sports) are picked.

In [ ]:
data = []

### **Web Scraping**

The next step is to scrape information from the news page. In this step, Beautiful Soup library is used to scrape information from each media

In [ ]:
# Kompas
for row in Kompas:
  r = requests.get(row[0])

  soup = BeautifulSoup(r.content, 'html5lib')

  article = soup.findAll('div', {'class': 'read__content'})

  paragraph = []
  for par in article:
    lines = par.findAll('p')
    for line in lines:
      # Exclude not needed elements
      if 'Baca juga' not in line.text and line.text.strip():
        a = line.find('a')
        if a:
          a.decompose()
        paragraph.append(line.text)

  row[0] = paragraph
  data.append(row)

In [ ]:
#CNN Indonesia
for row in CNN:
  r = requests.get(row[0])

  soup = BeautifulSoup(r.content, 'html5lib')

  article = soup.findAll('div', {'class': 'detail-text'})

  paragraph = []
  for par in article:
    lines = par.findAll('p')
    for line in lines:
      # Exclude not needed elements
      if 'ADVERTISEMENT' not in line.text and 'SCROLL TO CONTINUE WITH CONTENT' not in line.text and line.text.strip():
        paragraph.append(line.text)

  row[0] = paragraph
  data.append(row)

In [ ]:
#Detik
for row in Detik:
  r = requests.get(row[0])

  soup = BeautifulSoup(r.content, 'html5lib')

  article = soup.findAll('div', {'class': 'detail__body-text'})

  paragraph = []
  for par in article:
    lines = par.findAll('p')
    for line in lines:
      # Exclude not needed elements
      if 'ADVERTISEMENT' not in line.text and 'SCROLL TO CONTINUE WITH CONTENT' not in line.text and line.text.strip():
        paragraph.append(line.text)

  row[0] = paragraph
  data.append(row)

In [ ]:
#Liputan6
for row in Liputan6:
  r = requests.get(row[0])

  soup = BeautifulSoup(r.content, 'html5lib')

  article = soup.findAll('div', {'class': 'article-content-body__item-content'})

  paragraph = []
  for par in article:
    lines = par.findAll('p')
    for line in lines:
      # Exclude not needed elements
      if '* ' not in line.text and 'Baca Juga' not in line.text and 'Advertisement' not in line.text and line.text.strip():
        paragraph.append(line.text)

  row[0] = paragraph
  data.append(row)

In [ ]:
# Load the data into a dataframe

df = pd.DataFrame(data, columns=['Berita', 'Media', 'Label'])
df

Berita          Media     Label
0   [KOMPAS.com - Jakarta Coffee Week (Jacoweek) 2...         Kompas   Hiburan
1   [JAKARTA, KOMPAS.com - Wakil Presiden ke 10 da...         Kompas   Politik
2   [KOMPAS.com - Pemain ganda putra Indonesia,  (...         Kompas  Olahraga
3   [JAKARTA, KOMPAS.com - Komisi Pemberantasan Ko...         Kompas   Politik
4   [KOMPAS.com - Kepulan asap dari segelas kopi j...         Kompas   Hiburan
5   [KOMPAS.com - Dua wakil Indonesia tersingkir p...         Kompas  Olahraga
6   [JAKARTA, KOMPAS.com - Band legendaris  meneri...         Kompas   Hiburan
7   [KOMPAS.com - Timnas U17 Indonesia bermain imb...         Kompas  Olahraga
8   [Bioskop Trans TV malam ini, Sabtu (4/11), aka...  CNN Indonesia   Hiburan
9   [Sekretaris Jenderal (Sekjen) PDIP Hasto Krist...  CNN Indonesia   Politik
10  [Rehan Naufal Kusharjanto/Lisa Ayu Kusumawati ...  CNN Indonesia  Olahraga
11  [Alex Marquez jadi pembalap tercepat disusul J...  CNN Indonesia  Olahraga
12  [Jakarta, Indonesia menjadi salah satu kota ya...  CNN Indonesia   Hiburan
13  [Wakil Ketua DPR Lodewijk Freidrich Paulus ber...  CNN Indonesia   Politik
14  [Daftar lengkap nominasi Grammy Awards 2024 re...  CNN Indonesia   Hiburan
15  [Anggota Komisi I DPR Bobby Rizaldi mengatakan...  CNN Indonesia   Politik
16  [Harry Kane menginspirasi kemenangan Bayern Mu...          Detik  Olahraga
17  [Partai Gerindra menghormati keputusan PDIP ya...          Detik   Politik
18  [Kurang dari sebulan lagi Sweet Home 2 akan ta...          Detik   Hiburan
19  [San Antonio Spurs mengalahkan tuan rumah Phoe...          Detik  Olahraga
20  [Siapa sih yang tak kenal dengan Doraemon? Ser...          Detik   Hiburan
21  [Presiden Joko Widodo (Jokowi) akan menghadiri...          Detik   Politik
22  [Chelsea pelan-pelan menjauh dari papan bawah ...          Detik  Olahraga
23  [KPK menggeledah rumah Ketua Komisi IV DPR RI ...          Detik   Politik
24  [Liputan6.com, Jakarta - Taman hiburan Chimelo...       Liputan6   Hiburan
25  [Liputan6.com, Jakarta - Dash Sports, bekerja ...       Liputan6  Olahraga
26  [Liputan6.com, Jakarta - Puluhan mahasiswa di ...       Liputan6   Politik
27  [Liputan6.com, Jakarta Real Madrid dilaporkan ...       Liputan6  Olahraga
28  [Liputan6.com, Jakarta Drakor populer tahun 20...       Liputan6   Hiburan
29  [Liputan6.com, Jakarta Mantan Ketua Mahkamah K...       Liputan6   Politik
30  [Liputan6.com, Jakarta - Penyanyi Aruma kembal...       Liputan6   Hiburan
31  [Liputan6.com, Jakarta Los Angeles Clippers ke...       Liputan6  Olahraga

The dataframe above shows the result of web scraping. The 'Berita' feature contained raw text from scraping each urls and need to be preprocessed before used to build model.

The dataset consists of 32 entities and 3 features, with the features are:
*   Berita: News content (text data)
*   Media: Source of media (Kompas, CNN Indonesia, Detik, and Liputan6)
*   Label: News category (Politics, Entertainment, and Sports)

In [ ]:
# Save it to raw_data.csv as checkpoint

df.to_csv('raw_data.csv', index=False)

In [ ]:
raw_df = pd.read_csv('raw_data.csv')
raw_df

Berita          Media     Label
0   ['KOMPAS.com - Jakarta Coffee Week (Jacoweek) ...         Kompas   Hiburan
1   ['JAKARTA, KOMPAS.com - Wakil Presiden ke 10 d...         Kompas   Politik
2   ['KOMPAS.com - Pemain ganda putra Indonesia,  ...         Kompas  Olahraga
3   ['JAKARTA, KOMPAS.com - Komisi Pemberantasan K...         Kompas   Politik
4   ['KOMPAS.com -\xa0Kepulan asap dari segelas ko...         Kompas   Hiburan
5   ['KOMPAS.com - Dua wakil Indonesia tersingkir ...         Kompas  Olahraga
6   ['JAKARTA, KOMPAS.com - Band legendaris  mener...         Kompas   Hiburan
7   ['KOMPAS.com - Timnas U17 Indonesia bermain im...         Kompas  Olahraga
8   ['Bioskop Trans TV\xa0malam ini, Sabtu (4/11),...  CNN Indonesia   Hiburan
9   ['Sekretaris Jenderal (Sekjen) PDIP\xa0Hasto K...  CNN Indonesia   Politik
10  ['Rehan Naufal Kusharjanto/Lisa Ayu Kusumawati...  CNN Indonesia  Olahraga
11  ['Alex Marquez\xa0jadi pembalap tercepat disus...  CNN Indonesia  Olahraga
12  ['Jakarta, Indonesia menjadi salah satu kota y...  CNN Indonesia   Hiburan
13  ['Wakil Ketua DPR\xa0Lodewijk Freidrich Paulus...  CNN Indonesia   Politik
14  ['Daftar lengkap nominasi\xa0Grammy Awards 202...  CNN Indonesia   Hiburan
15  ['Anggota\xa0Komisi I DPR Bobby Rizaldi mengat...  CNN Indonesia   Politik
16  ['Harry Kane menginspirasi kemenangan Bayern M...          Detik  Olahraga
17  ['Partai Gerindra menghormati keputusan PDIP y...          Detik   Politik
18  ['Kurang dari sebulan lagi Sweet Home 2 akan t...          Detik   Hiburan
19  ['San Antonio Spurs mengalahkan tuan rumah Pho...          Detik  Olahraga
20  ['Siapa sih yang tak kenal dengan Doraemon? Se...          Detik   Hiburan
21  ['Presiden Joko Widodo (Jokowi) akan menghadir...          Detik   Politik
22  ['Chelsea pelan-pelan menjauh dari papan bawah...          Detik  Olahraga
23  ['KPK menggeledah rumah Ketua Komisi IV DPR RI...          Detik   Politik
24  ['Liputan6.com, Jakarta - Taman hiburan Chimel...       Liputan6   Hiburan
25  ['Liputan6.com, Jakarta - Dash Sports, bekerja...       Liputan6  Olahraga
26  ['Liputan6.com, Jakarta - Puluhan mahasiswa di...       Liputan6   Politik
27  ['Liputan6.com, Jakarta Real Madrid dilaporkan...       Liputan6  Olahraga
28  ['Liputan6.com, Jakarta Drakor populer tahun 2...       Liputan6   Hiburan
29  ['Liputan6.com, Jakarta Mantan Ketua Mahkamah ...       Liputan6   Politik
30  ['Liputan6.com, Jakarta - Penyanyi Aruma kemba...       Liputan6   Hiburan
31  ['Liputan6.com, Jakarta Los Angeles Clippers k...       Liputan6  Olahraga

## **Text Preprocessing**

### **Text Cleaning**

The first step of text preprocessing is text cleaning. In this process, the raw text will be cleaned. This process includes lowering text, removing punctuation, symbols, extra spaces, digits, urls, and unnecessary words.

In [ ]:
import re

In [ ]:
def clean_text(df, text_field):
  df[text_field] = df[text_field].str.lower()
  df[text_field] = df[text_field].apply(lambda elem: re.sub(r'xa0', ' ', elem))
  df[text_field] = df[text_field].apply(lambda elem: re.sub(r'(http[s]?\://\S+)|(#\S+)|(@\S+)|kompas\S+|liputan6\S+|baca juga\S+|\d+|(\n)+|(\[[A-Za-z0-9]+)|([^0-9A-Za-z \t])|(\w+:\/\/\S+)', '', elem))
  df[text_field] = df[text_field].apply(lambda elem: re.sub(r'(\s){2,}', ' ', elem))
  df[text_field] = df[text_field].apply(lambda elem: re.sub(r'^(\s+)|(\s+)$', '', elem))

  return df

In [ ]:
clean_df = clean_text(raw_df, "Berita")
clean_df

Berita          Media     Label
0   jakarta coffee week jacoweek akan kembali dige...         Kompas   Hiburan
1   jakarta wakil presiden ke dan ri meminta pemer...         Kompas   Politik
2   pemain ganda putra indonesia tahun rela jika t...         Kompas  Olahraga
3   jakarta komisi pemberantasan korupsi menyataka...         Kompas   Politik
4   kepulan asap dari segelas kopi jos tidak hanya...         Kompas   Hiburan
5   dua wakil indonesia tersingkir pada babak besa...         Kompas  Olahraga
6   jakarta band legendaris menerima rekor muri se...         Kompas   Hiburan
7   timnas u indonesia bermain imbang dengan pada ...         Kompas  Olahraga
8   bioskop trans tv malam ini sabtu akan menayang...  CNN Indonesia   Hiburan
9   sekretaris jenderal sekjen pdip hasto kristiya...  CNN Indonesia   Politik
10  rehan naufal kusharjantolisa ayu kusumawati lo...  CNN Indonesia  Olahraga
11  alex marquez jadi pembalap tercepat disusul jo...  CNN Indonesia  Olahraga
12  jakarta indonesia menjadi salah satu kota yang...  CNN Indonesia   Hiburan
13  wakil ketua dpr lodewijk freidrich paulus berh...  CNN Indonesia   Politik
14  daftar lengkap nominasi grammy awards resmi di...  CNN Indonesia   Hiburan
15  anggota komisi i dpr bobby rizaldi mengatakan ...  CNN Indonesia   Politik
16  harry kane menginspirasi kemenangan bayern mun...          Detik  Olahraga
17  partai gerindra menghormati keputusan pdip yan...          Detik   Politik
18  kurang dari sebulan lagi sweet home akan tayan...          Detik   Hiburan
19  san antonio spurs mengalahkan tuan rumah phoen...          Detik  Olahraga
20  siapa sih yang tak kenal dengan doraemon seria...          Detik   Hiburan
21  presiden joko widodo jokowi akan menghadiri ko...          Detik   Politik
22  chelsea pelanpelan menjauh dari papan bawah li...          Detik  Olahraga
23  kpk menggeledah rumah ketua komisi iv dpr ri d...          Detik   Politik
24  jakarta taman hiburan chimelong yang merupakan...       Liputan6   Hiburan
25  jakarta dash sports bekerja sama dengan persat...       Liputan6  Olahraga
26  jakarta puluhan mahasiswa di solo jawa tengah ...       Liputan6   Politik
27  jakarta real madrid dilaporkan sudah memilih p...       Liputan6  Olahraga
28  jakarta drakor populer tahun descendants of th...       Liputan6   Hiburan
29  jakarta mantan ketua mahkamah konstitusi mk an...       Liputan6   Politik
30  jakarta penyanyi aruma kembali menorehkan pres...       Liputan6   Hiburan
31  jakarta los angeles clippers kembali menelan k...       Liputan6  Olahraga

In [ ]:
# Save it to cleaned_data.csv as checkpoint
clean_df.to_csv('clean_data.csv', index=False)

In [ ]:
cleaned_df = pd.read_csv('clean_data.csv')
cleaned_df

Berita          Media     Label
0   jakarta coffee week jacoweek akan kembali dige...         Kompas   Hiburan
1   jakarta wakil presiden ke dan ri meminta pemer...         Kompas   Politik
2   pemain ganda putra indonesia tahun rela jika t...         Kompas  Olahraga
3   jakarta komisi pemberantasan korupsi menyataka...         Kompas   Politik
4   kepulan asap dari segelas kopi jos tidak hanya...         Kompas   Hiburan
5   dua wakil indonesia tersingkir pada babak besa...         Kompas  Olahraga
6   jakarta band legendaris menerima rekor muri se...         Kompas   Hiburan
7   timnas u indonesia bermain imbang dengan pada ...         Kompas  Olahraga
8   bioskop trans tv malam ini sabtu akan menayang...  CNN Indonesia   Hiburan
9   sekretaris jenderal sekjen pdip hasto kristiya...  CNN Indonesia   Politik
10  rehan naufal kusharjantolisa ayu kusumawati lo...  CNN Indonesia  Olahraga
11  alex marquez jadi pembalap tercepat disusul jo...  CNN Indonesia  Olahraga
12  jakarta indonesia menjadi salah satu kota yang...  CNN Indonesia   Hiburan
13  wakil ketua dpr lodewijk freidrich paulus berh...  CNN Indonesia   Politik
14  daftar lengkap nominasi grammy awards resmi di...  CNN Indonesia   Hiburan
15  anggota komisi i dpr bobby rizaldi mengatakan ...  CNN Indonesia   Politik
16  harry kane menginspirasi kemenangan bayern mun...          Detik  Olahraga
17  partai gerindra menghormati keputusan pdip yan...          Detik   Politik
18  kurang dari sebulan lagi sweet home akan tayan...          Detik   Hiburan
19  san antonio spurs mengalahkan tuan rumah phoen...          Detik  Olahraga
20  siapa sih yang tak kenal dengan doraemon seria...          Detik   Hiburan
21  presiden joko widodo jokowi akan menghadiri ko...          Detik   Politik
22  chelsea pelanpelan menjauh dari papan bawah li...          Detik  Olahraga
23  kpk menggeledah rumah ketua komisi iv dpr ri d...          Detik   Politik
24  jakarta taman hiburan chimelong yang merupakan...       Liputan6   Hiburan
25  jakarta dash sports bekerja sama dengan persat...       Liputan6  Olahraga
26  jakarta puluhan mahasiswa di solo jawa tengah ...       Liputan6   Politik
27  jakarta real madrid dilaporkan sudah memilih p...       Liputan6  Olahraga
28  jakarta drakor populer tahun descendants of th...       Liputan6   Hiburan
29  jakarta mantan ketua mahkamah konstitusi mk an...       Liputan6   Politik
30  jakarta penyanyi aruma kembali menorehkan pres...       Liputan6   Hiburan
31  jakarta los angeles clippers kembali menelan k...       Liputan6  Olahraga

### **Tokenization**

The next step is to tokenize the paragraph/text into smaller units, which is word. This process is done with the help of NLTK library.

In [ ]:
filtered_df = cleaned_df

In [ ]:
# Import libraries
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize

# Function to perform tokenization
def tokenize_text(text):
  tokens = nltk.tokenize.word_tokenize(text)
  return tokens

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
filtered_df['Berita'] = filtered_df['Berita'].apply(tokenize_text)

In [ ]:
filtered_df

Berita          Media     Label
0   [jakarta, coffee, week, jacoweek, akan, kembal...         Kompas   Hiburan
1   [jakarta, wakil, presiden, ke, dan, ri, memint...         Kompas   Politik
2   [pemain, ganda, putra, indonesia, tahun, rela,...         Kompas  Olahraga
3   [jakarta, komisi, pemberantasan, korupsi, meny...         Kompas   Politik
4   [kepulan, asap, dari, segelas, kopi, jos, tida...         Kompas   Hiburan
5   [dua, wakil, indonesia, tersingkir, pada, baba...         Kompas  Olahraga
6   [jakarta, band, legendaris, menerima, rekor, m...         Kompas   Hiburan
7   [timnas, u, indonesia, bermain, imbang, dengan...         Kompas  Olahraga
8   [bioskop, trans, tv, malam, ini, sabtu, akan, ...  CNN Indonesia   Hiburan
9   [sekretaris, jenderal, sekjen, pdip, hasto, kr...  CNN Indonesia   Politik
10  [rehan, naufal, kusharjantolisa, ayu, kusumawa...  CNN Indonesia  Olahraga
11  [alex, marquez, jadi, pembalap, tercepat, disu...  CNN Indonesia  Olahraga
12  [jakarta, indonesia, menjadi, salah, satu, kot...  CNN Indonesia   Hiburan
13  [wakil, ketua, dpr, lodewijk, freidrich, paulu...  CNN Indonesia   Politik
14  [daftar, lengkap, nominasi, grammy, awards, re...  CNN Indonesia   Hiburan
15  [anggota, komisi, i, dpr, bobby, rizaldi, meng...  CNN Indonesia   Politik
16  [harry, kane, menginspirasi, kemenangan, bayer...          Detik  Olahraga
17  [partai, gerindra, menghormati, keputusan, pdi...          Detik   Politik
18  [kurang, dari, sebulan, lagi, sweet, home, aka...          Detik   Hiburan
19  [san, antonio, spurs, mengalahkan, tuan, rumah...          Detik  Olahraga
20  [siapa, sih, yang, tak, kenal, dengan, doraemo...          Detik   Hiburan
21  [presiden, joko, widodo, jokowi, akan, menghad...          Detik   Politik
22  [chelsea, pelanpelan, menjauh, dari, papan, ba...          Detik  Olahraga
23  [kpk, menggeledah, rumah, ketua, komisi, iv, d...          Detik   Politik
24  [jakarta, taman, hiburan, chimelong, yang, mer...       Liputan6   Hiburan
25  [jakarta, dash, sports, bekerja, sama, dengan,...       Liputan6  Olahraga
26  [jakarta, puluhan, mahasiswa, di, solo, jawa, ...       Liputan6   Politik
27  [jakarta, real, madrid, dilaporkan, sudah, mem...       Liputan6  Olahraga
28  [jakarta, drakor, populer, tahun, descendants,...       Liputan6   Hiburan
29  [jakarta, mantan, ketua, mahkamah, konstitusi,...       Liputan6   Politik
30  [jakarta, penyanyi, aruma, kembali, menorehkan...       Liputan6   Hiburan
31  [jakarta, los, angeles, clippers, kembali, men...       Liputan6  Olahraga

### **Stemming**

It is a text normalization process to reduce words to their base form by removing suffixes or prefixes from words. This process is done with the help of Sastrawi library, which is a library that contains words in Bahasa Indonesia.

In [ ]:
!pip install Sastrawi

In [ ]:
# Import library
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

factory = StemmerFactory()
stemmer = factory.create_stemmer()

In [ ]:
# Function to perform token stemming
def stem_tokens(tokens):
    stemmed_tokens = [stemmer.stem(token) for token in tokens]
    return stemmed_tokens

In [ ]:
filtered_df['Berita'] = filtered_df['Berita'].apply(stem_tokens)

In [ ]:
filtered_df

Berita          Media     Label
0   [jakarta, coffee, week, jacoweek, akan, kembal...         Kompas   Hiburan
1   [jakarta, wakil, presiden, ke, dan, ri, minta,...         Kompas   Politik
2   [main, ganda, putra, indonesia, tahun, rela, j...         Kompas  Olahraga
3   [jakarta, komisi, berantas, korupsi, nyata, te...         Kompas   Politik
4   [kepul, asap, dari, gelas, kopi, jos, tidak, h...         Kompas   Hiburan
5   [dua, wakil, indonesia, singkir, pada, babak, ...         Kompas  Olahraga
6   [jakarta, band, legendaris, terima, rekor, mur...         Kompas   Hiburan
7   [timnas, u, indonesia, main, imbang, dengan, p...         Kompas  Olahraga
8   [bioskop, trans, tv, malam, ini, sabtu, akan, ...  CNN Indonesia   Hiburan
9   [sekretaris, jenderal, sekjen, pdip, hasto, kr...  CNN Indonesia   Politik
10  [rehan, naufal, kusharjantolisa, ayu, kusumawa...  CNN Indonesia  Olahraga
11  [alex, marquez, jadi, balap, cepat, susul, jor...  CNN Indonesia  Olahraga
12  [jakarta, indonesia, jadi, salah, satu, kota, ...  CNN Indonesia   Hiburan
13  [wakil, ketua, dpr, lodewijk, freidrich, paulu...  CNN Indonesia   Politik
14  [daftar, lengkap, nominasi, grammy, awards, re...  CNN Indonesia   Hiburan
15  [anggota, komisi, i, dpr, bobby, rizaldi, kata...  CNN Indonesia   Politik
16  [harry, kane, inspirasi, menang, bayern, munic...          Detik  Olahraga
17  [partai, gerindra, hormat, putus, pdip, yang, ...          Detik   Politik
18  [kurang, dari, bulan, lagi, sweet, home, akan,...          Detik   Hiburan
19  [san, antonio, spurs, kalah, tuan, rumah, phoe...          Detik  Olahraga
20  [siapa, sih, yang, tak, kenal, dengan, doraemo...          Detik   Hiburan
21  [presiden, joko, widodo, jokowi, akan, hadir, ...          Detik   Politik
22  [chelsea, pelanpelan, jauh, dari, papan, bawah...          Detik  Olahraga
23  [kpk, geledah, rumah, ketua, komisi, iv, dpr, ...          Detik   Politik
24  [jakarta, taman, hibur, chimelong, yang, rupa,...       Liputan6   Hiburan
25  [jakarta, dash, sports, kerja, sama, dengan, s...       Liputan6  Olahraga
26  [jakarta, puluh, mahasiswa, di, solo, jawa, te...       Liputan6   Politik
27  [jakarta, real, madrid, lapor, sudah, pilih, m...       Liputan6  Olahraga
28  [jakarta, drakor, populer, tahun, descendants,...       Liputan6   Hiburan
29  [jakarta, mantan, ketua, mahkamah, konstitusi,...       Liputan6   Politik
30  [jakarta, nyanyi, aruma, kembali, toreh, prest...       Liputan6   Hiburan
31  [jakarta, los, angeles, clippers, kembali, tel...       Liputan6  Olahraga

### **Removing Stop Words**

This process aims to filter out stop words as they considered to be of little value in text analysis. This process is done with the help of NLTK library + additional custom Indonesia stop words found from github.

In [ ]:
# Custom Indonesian Stop Words

with open('id.stopwords.02.01.2016.txt') as f:
  custom_stop = f.read().splitlines()

In [ ]:
# Remove Stop Words
nltk.download('stopwords')
from nltk.corpus import stopwords

# Stop Words NLTK extended with custom stop words
stopwords = stopwords.words('indonesian')
stopwords.clear()
stopwords.extend(custom_stop)
stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['ada',
 'adalah',
 'adanya',
 'adapun',
 'agak',
 'agaknya',
 'agar',
 'akan',
 'akankah',
 'akhir',
 'akhiri',
 'akhirnya',
 'aku',
 'akulah',
 'amat',
 'amatlah',
 'anda',
 'andalah',
 'antar',
 'antara',
 'antaranya',
 'apa',
 'apaan',
 'apabila',
 'apakah',
 'apalagi',
 'apatah',
 'artinya',
 'asal',
 'asalkan',
 'atas',
 'atau',
 'ataukah',
 'ataupun',
 'awal',
 'awalnya',
 'bagai',
 'bagaikan',
 'bagaimana',
 'bagaimanakah',
 'bagaimanapun',
 'bagi',
 'bagian',
 'bahkan',
 'bahwa',
 'bahwasanya',
 'baik',
 'bakal',
 'bakalan',
 'balik',
 'banyak',
 'bapak',
 'baru',
 'bawah',
 'beberapa',
 'begini',
 'beginian',
 'beginikah',
 'beginilah',
 'begitu',
 'begitukah',
 'begitulah',
 'begitupun',
 'bekerja',
 'belakang',
 'belakangan',
 'belum',
 'belumlah',
 'benar',
 'benarkah',
 'benarlah',
 'berada',
 'berakhir',
 'berakhirlah',
 'berakhirnya',
 'berapa',
 'berapakah',
 'berapalah',
 'berapapun',
 'berarti',
 'berawal',
 'berbagai',
 'berdatangan',
 'beri',
 'berikan',
 'berikut'

In [ ]:
# Function to perform stop words removal
def remove_stopwords(tokens):
    filtered_tokens = [word for word in tokens if word not in stopwords]
    return filtered_tokens

In [ ]:
filtered_df['Berita'] = filtered_df['Berita'].apply(remove_stopwords)

In [ ]:
filtered_df

Berita          Media     Label
0   [jakarta, coffee, week, jacoweek, gelar, novem...         Kompas   Hiburan
1   [jakarta, wakil, presiden, ri, perintah, indon...         Kompas   Politik
2   [main, ganda, putra, indonesia, rela, pasang, ...         Kompas  Olahraga
3   [jakarta, komisi, berantas, korupsi, nyata, sa...         Kompas   Politik
4   [kepul, asap, gelas, kopi, jos, air, panas, ar...         Kompas   Hiburan
5   [wakil, indonesia, singkir, babak, adnan, maul...         Kompas  Olahraga
6   [jakarta, band, legendaris, terima, rekor, mur...         Kompas   Hiburan
7   [timnas, u, indonesia, main, imbang, laga, gru...         Kompas  Olahraga
8   [bioskop, trans, tv, malam, sabtu, tayang, son...  CNN Indonesia   Hiburan
9   [sekretaris, jenderal, sekjen, pdip, hasto, kr...  CNN Indonesia   Politik
10  [rehan, naufal, kusharjantolisa, ayu, kusumawa...  CNN Indonesia  Olahraga
11  [alex, marquez, balap, cepat, susul, jorge, ma...  CNN Indonesia  Olahraga
12  [jakarta, indonesia, salah, kota, sambang, nct...  CNN Indonesia   Hiburan
13  [wakil, ketua, dpr, lodewijk, freidrich, paulu...  CNN Indonesia   Politik
14  [daftar, lengkap, nominasi, grammy, awards, re...  CNN Indonesia   Hiburan
15  [anggota, komisi, i, dpr, bobby, rizaldi, milu...  CNN Indonesia   Politik
16  [harry, kane, inspirasi, menang, bayern, munic...          Detik  Olahraga
17  [partai, gerindra, hormat, putus, pdip, tarik,...          Detik   Politik
18  [sweet, home, tayang, netflix, teaser, trailer...          Detik   Hiburan
19  [san, antonio, spurs, kalah, tuan, rumah, phoe...          Detik  Olahraga
20  [sih, kenal, doraemon, serial, komik, anime, f...          Detik   Hiburan
21  [presiden, joko, widodo, jokowi, hadir, konfer...          Detik   Politik
22  [chelsea, pelanpelan, papan, liga, inggris, ma...          Detik  Olahraga
23  [kpk, geledah, rumah, ketua, komisi, iv, dpr, ...          Detik   Politik
24  [jakarta, taman, hibur, chimelong, rupa, buah,...       Liputan6   Hiburan
25  [jakarta, dash, sports, kerja, bola, basket, i...       Liputan6  Olahraga
26  [jakarta, puluh, mahasiswa, solo, jawa, jateng...       Liputan6   Politik
27  [jakarta, real, madrid, lapor, pilih, main, re...       Liputan6  Olahraga
28  [jakarta, drakor, populer, descendants, of, th...       Liputan6   Hiburan
29  [jakarta, mantan, ketua, mahkamah, konstitusi,...       Liputan6   Politik
30  [jakarta, nyanyi, aruma, toreh, prestasi, gemi...       Liputan6   Hiburan
31  [jakarta, los, angeles, clippers, tel, kalah, ...       Liputan6  Olahraga

In [ ]:
filtered_df['Berita'] = filtered_df['Berita'].apply(lambda x: ' '.join(x))

In [ ]:
filtered_df

Berita          Media     Label
0   jakarta coffee week jacoweek gelar november in...         Kompas   Hiburan
1   jakarta wakil presiden ri perintah indonesia r...         Kompas   Politik
2   main ganda putra indonesia rela pasang kevin s...         Kompas  Olahraga
3   jakarta komisi berantas korupsi nyata sangka d...         Kompas   Politik
4   kepul asap gelas kopi jos air panas arang beda...         Kompas   Hiburan
5   wakil indonesia singkir babak adnan maulananit...         Kompas  Olahraga
6   jakarta band legendaris terima rekor muri acar...         Kompas   Hiburan
7   timnas u indonesia main imbang laga grup a pia...         Kompas  Olahraga
8   bioskop trans tv malam sabtu tayang songbird k...  CNN Indonesia   Hiburan
9   sekretaris jenderal sekjen pdip hasto kristiya...  CNN Indonesia   Politik
10  rehan naufal kusharjantolisa ayu kusumawati lo...  CNN Indonesia  Olahraga
11  alex marquez balap cepat susul jorge martin po...  CNN Indonesia  Olahraga
12  jakarta indonesia salah kota sambang nct tur k...  CNN Indonesia   Hiburan
13  wakil ketua dpr lodewijk freidrich paulus hara...  CNN Indonesia   Politik
14  daftar lengkap nominasi grammy awards resmi re...  CNN Indonesia   Hiburan
15  anggota komisi i dpr bobby rizaldi milu papua ...  CNN Indonesia   Politik
16  harry kane inspirasi menang bayern munich der ...          Detik  Olahraga
17  partai gerindra hormat putus pdip tarik menter...          Detik   Politik
18  sweet home tayang netflix teaser trailer rilis...          Detik   Hiburan
19  san antonio spurs kalah tuan rumah phoenix sun...          Detik  Olahraga
20  sih kenal doraemon serial komik anime fujiko f...          Detik   Hiburan
21  presiden joko widodo jokowi hadir konferensi t...          Detik   Politik
22  chelsea pelanpelan papan liga inggris main tam...          Detik  Olahraga
23  kpk geledah rumah ketua komisi iv dpr ri fraks...          Detik   Politik
24  jakarta taman hibur chimelong rupa buah taman ...       Liputan6   Hiburan
25  jakarta dash sports kerja bola basket indonesi...       Liputan6  Olahraga
26  jakarta puluh mahasiswa solo jawa jateng gabun...       Liputan6   Politik
27  jakarta real madrid lapor pilih main rekrut ky...       Liputan6  Olahraga
28  jakarta drakor populer descendants of the sun ...       Liputan6   Hiburan
29  jakarta mantan ketua mahkamah konstitusi mk an...       Liputan6   Politik
30  jakarta nyanyi aruma toreh prestasi gemilang k...       Liputan6   Hiburan
31  jakarta los angeles clippers tel kalah main ja...       Liputan6  Olahraga

In [ ]:
# Save it to filtered_data.csv as checkpoint
filtered_df.to_csv('filtered_data.csv', index=False)

In [ ]:
data = pd.read_csv('filtered_data.csv')

In [ ]:
data

Berita          Media     Label
0   jakarta coffee week jacoweek gelar november in...         Kompas   Hiburan
1   jakarta wakil presiden ri perintah indonesia r...         Kompas   Politik
2   main ganda putra indonesia rela pasang kevin s...         Kompas  Olahraga
3   jakarta komisi berantas korupsi nyata sangka d...         Kompas   Politik
4   kepul asap gelas kopi jos air panas arang beda...         Kompas   Hiburan
5   wakil indonesia singkir babak adnan maulananit...         Kompas  Olahraga
6   jakarta band legendaris terima rekor muri acar...         Kompas   Hiburan
7   timnas u indonesia main imbang laga grup a pia...         Kompas  Olahraga
8   bioskop trans tv malam sabtu tayang songbird k...  CNN Indonesia   Hiburan
9   sekretaris jenderal sekjen pdip hasto kristiya...  CNN Indonesia   Politik
10  rehan naufal kusharjantolisa ayu kusumawati lo...  CNN Indonesia  Olahraga
11  alex marquez balap cepat susul jorge martin po...  CNN Indonesia  Olahraga
12  jakarta indonesia salah kota sambang nct tur k...  CNN Indonesia   Hiburan
13  wakil ketua dpr lodewijk freidrich paulus hara...  CNN Indonesia   Politik
14  daftar lengkap nominasi grammy awards resmi re...  CNN Indonesia   Hiburan
15  anggota komisi i dpr bobby rizaldi milu papua ...  CNN Indonesia   Politik
16  harry kane inspirasi menang bayern munich der ...          Detik  Olahraga
17  partai gerindra hormat putus pdip tarik menter...          Detik   Politik
18  sweet home tayang netflix teaser trailer rilis...          Detik   Hiburan
19  san antonio spurs kalah tuan rumah phoenix sun...          Detik  Olahraga
20  sih kenal doraemon serial komik anime fujiko f...          Detik   Hiburan
21  presiden joko widodo jokowi hadir konferensi t...          Detik   Politik
22  chelsea pelanpelan papan liga inggris main tam...          Detik  Olahraga
23  kpk geledah rumah ketua komisi iv dpr ri fraks...          Detik   Politik
24  jakarta taman hibur chimelong rupa buah taman ...       Liputan6   Hiburan
25  jakarta dash sports kerja bola basket indonesi...       Liputan6  Olahraga
26  jakarta puluh mahasiswa solo jawa jateng gabun...       Liputan6   Politik
27  jakarta real madrid lapor pilih main rekrut ky...       Liputan6  Olahraga
28  jakarta drakor populer descendants of the sun ...       Liputan6   Hiburan
29  jakarta mantan ketua mahkamah konstitusi mk an...       Liputan6   Politik
30  jakarta nyanyi aruma toreh prestasi gemilang k...       Liputan6   Hiburan
31  jakarta los angeles clippers tel kalah main ja...       Liputan6  Olahraga

In [ ]:
# Import Libraries
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, classification_report, confusion_matrix

from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer

## **Model and Evaluation**

The machine learning model is built using two algorithms: Support Vector Machine (SVM) and Random Forest. The two algorithm will be evaluated by comparing several aspects: precision, recall, f1-score, and support.

### **Text Representation**

Before the model is built, text representation is performed. It is a process of converting textual data into a numerical format that can be understood and processed by machine learning algorithms and models. The process is done using two methods: TfidVectorizer and CountVectorizer.

#### **TfidfVectorizer**

In [ ]:
# TfidfVectorizer

# Initialize a TfidfVectorizer object with vector size: 50 and min frequency: 3
tfidf = TfidfVectorizer(max_features=50, min_df=3)

tfidf_matrix = tfidf.fit_transform(data['Berita'])

tfidf_df = pd.DataFrame(data=tfidf_matrix.toarray(), columns=tfidf.get_feature_names_out())
tfidf_df.head()

alex  and  bola   cerita  detik  dpr     dunia     gelar  gibran  \
0  0.000000  0.0   0.0  0.00000    0.0  0.0  0.140647  0.154095     0.0   
1  0.000000  0.0   0.0  0.00000    0.0  0.0  0.000000  0.000000     0.0   
2  0.000000  0.0   0.0  0.00000    0.0  0.0  0.000000  0.000000     0.0   
3  0.738002  0.0   0.0  0.00000    0.0  0.0  0.000000  0.000000     0.0   
4  0.000000  0.0   0.0  0.41756    0.0  0.0  0.000000  0.000000     0.0   

      hadir  ...  putus  salah  skor  tampil  tanding  terima       the  ubah  \
0  0.362538  ...    0.0    0.0   0.0     0.0      0.0     0.0  0.170938   0.0   
1  0.000000  ...    0.0    0.0   0.0     0.0      0.0     0.0  0.000000   0.0   
2  0.000000  ...    0.0    0.0   0.0     0.0      0.0     0.0  0.000000   0.0   
3  0.000000  ...    0.0    0.0   0.0     0.0      0.0     0.0  0.000000   0.0   
4  0.000000  ...    0.0    0.0   0.0     0.0      0.0     0.0  0.000000   0.0   

   unggul     wakil  
0     0.0  0.000000  
1     0.0  0.165336  
2     0.0  0.000000  
3     0.0  0.205953  
4     0.0  0.000000  

[5 rows x 50 columns]

In [ ]:
# Train test split

X_tfidf = tfidf_df
y_tfidf = data['Label']
X_train_tfidf, X_test_tfidf, y_train_tfidf, y_test_tfidf = train_test_split(X_tfidf, y_tfidf, test_size=0.2, random_state=36)

In [ ]:
print("Training Data:")
print(X_train_tfidf.head())
print(y_train_tfidf.head())

print("Test Data:")
print(X_test_tfidf.head())
print(y_test_tfidf.head())

Training Data:
        alex  and  bola  cerita     detik       dpr  dunia  gelar  gibran  \
29  0.000000  0.0   0.0     0.0  0.000000  0.047786    0.0    0.0     0.0   
23  0.178929  0.0   0.0     0.0  0.178929  0.715716    0.0    0.0     0.0   
16  0.000000  0.0   0.0     0.0  0.000000  0.000000    0.0    0.0     0.0   
11  0.348694  0.0   0.0     0.0  0.906605  0.000000    0.0    0.0     0.0   
3   0.738002  0.0   0.0     0.0  0.000000  0.000000    0.0    0.0     0.0   

    hadir  ...     putus     salah      skor    tampil   tanding  terima  the  \
29    0.0  ...  0.295517  0.000000  0.000000  0.000000  0.000000     0.0  0.0   
23    0.0  ...  0.000000  0.135971  0.000000  0.000000  0.000000     0.0  0.0   
16    0.0  ...  0.000000  0.000000  0.191217  0.172376  0.608321     0.0  0.0   
11    0.0  ...  0.000000  0.000000  0.000000  0.111081  0.000000     0.0  0.0   
3     0.0  ...  0.000000  0.000000  0.000000  0.000000  0.000000     0.0  0.0   

        ubah    unggul     wakil  


In [ ]:
# Support Vector Machine
tfidf_svm = SVC()

tfidf_svm.fit(X_train_tfidf, y_train_tfidf)

tfidf_svm_predict = tfidf_svm.predict(X_test_tfidf)

tfidf_svm_accuracy = accuracy_score(y_test_tfidf, tfidf_svm_predict)

print(tfidf_svm_accuracy)
print(classification_report(y_test_tfidf, tfidf_svm_predict))

1.0
              precision    recall  f1-score   support

     Hiburan       1.00      1.00      1.00         2
    Olahraga       1.00      1.00      1.00         2
     Politik       1.00      1.00      1.00         3

    accuracy                           1.00         7
   macro avg       1.00      1.00      1.00         7
weighted avg       1.00      1.00      1.00         7



The classification report above shows that SVM give 100% in all metrics for predicting all labels. This could indicates that there might be an issue with the data or the way the model is evaluated as it is quite rare to get 100% accuracy. This problem could occur due to small number of data is used.

In [ ]:
# Random Forest
tfidf_rf = RandomForestClassifier()

tfidf_rf.fit(X_train_tfidf, y_train_tfidf)

tfidf_rf_predict = tfidf_rf.predict(X_test_tfidf)

tfidf_rf_accuracy = accuracy_score(y_test_tfidf, tfidf_rf_predict)

print(tfidf_rf_accuracy)
print(classification_report(y_test_tfidf, tfidf_rf_predict))

0.5714285714285714
              precision    recall  f1-score   support

     Hiburan       0.33      0.50      0.40         2
    Olahraga       0.50      0.50      0.50         2
     Politik       1.00      0.67      0.80         3

    accuracy                           0.57         7
   macro avg       0.61      0.56      0.57         7
weighted avg       0.67      0.57      0.60         7



The classification report above shows that Random Forest give a high number on the 'Politics' metrics, medium number on the 'Entertainment' metrics, and small number on the 'Sports' metrics. And overall it gives 57% accuracy. This report could shows that the 'Entertainment' is underfitted. It could occur due to contents found in the 'Entertainment' is more variative so the model need to learn a bigger data.

#### **CountVectorizer**

In [ ]:
# CountVectorizer

# # Initialize a CountVectorizer object with vector size: 50 and min frequency: 3
cv = CountVectorizer(max_features=50, min_df=3)

cv_matrix = cv.fit_transform(data['Berita'])

cv_df = pd.DataFrame(data=cv_matrix.toarray(), columns=cv.get_feature_names_out())
cv_df.head()

alex  and  bola  cerita  detik  dpr  dunia  gelar  gibran  hadir  ...  \
0     0    0     0       0      0    0      1      1       0      2  ...   
1     0    0     0       0      0    0      0      0       0      0  ...   
2     0    0     0       0      0    0      0      0       0      0  ...   
3     3    0     0       0      0    0      0      0       0      0  ...   
4     0    0     0       1      0    0      0      0       0      0  ...   

   putus  salah  skor  tampil  tanding  terima  the  ubah  unggul  wakil  
0      0      0     0       0        0       0    1     0       0      0  
1      0      0     0       0        0       0    0     0       0      1  
2      0      0     0       0        0       0    0     0       0      0  
3      0      0     0       0        0       0    0     0       0      1  
4      0      0     0       0        0       0    0     0       0      0  

[5 rows x 50 columns]

In [ ]:
# Train test split
X_cv = cv_df
y_cv = data['Label']
X_train_cv, X_test_cv, y_train_cv, y_test_cv = train_test_split(X_cv, y_cv, test_size=0.2, random_state=36)

In [ ]:
# Support Vector Machine
cv_svm = SVC()

cv_svm.fit(X_train_cv, y_train_cv)

cv_svm_predict = cv_svm.predict(X_test_cv)

cv_svm_accuracy = accuracy_score(y_test_cv, cv_svm_predict)

print(cv_svm_accuracy)
print(classification_report(y_test_cv, cv_svm_predict))

0.7142857142857143
              precision    recall  f1-score   support

     Hiburan       0.50      1.00      0.67         2
    Olahraga       1.00      1.00      1.00         2
     Politik       1.00      0.33      0.50         3

    accuracy                           0.71         7
   macro avg       0.83      0.78      0.72         7
weighted avg       0.86      0.71      0.69         7



The classification report above shows that SVM gives an overall pretty good accuracy (71%). But we also can see that the 'Politics' recall metrics is so low (33%) and it tells that the model can classify 'Politics' well but still missing a lot of detecting the 'Politics' category.

In [ ]:
# Random Forest
cv_rf = RandomForestClassifier()

cv_rf.fit(X_train_cv, y_train_cv)

cv_rf_predict = cv_rf.predict(X_test_cv)

cv_rf_accuracy = accuracy_score(y_test_cv, cv_rf_predict)

print(cv_rf_accuracy)
print(classification_report(y_test_cv, cv_rf_predict))

0.8571428571428571
              precision    recall  f1-score   support

     Hiburan       0.67      1.00      0.80         2
    Olahraga       1.00      0.50      0.67         2
     Politik       1.00      1.00      1.00         3

    accuracy                           0.86         7
   macro avg       0.89      0.83      0.82         7
weighted avg       0.90      0.86      0.85         7



The classification report above shows that Random Forest gives an overall high accuracy (86%). The classification report metrics also shows a high number and it indicated that the Random Forest model performs pretty well. The lowest number is on the 'Olahraga' recall metrics and it indicates that the model is still missing a lot of detecting the 'Sports' category.

## **Analysis and Conclusion**

**Accuracy**

***TfidfVectorizer:***

SVM: 100%

Random Forest: 57.1%

***CountVectorizer:***

SVM: 71.4%

Random Forest: 85.7%


From the result above, we could see that using text representation TfidfVectorizer, the SVM performs better than Random Forest as it give 100% accuracy. Meanwhile, using CountVectorizer, the Random Forest performs better as it give 85.7% accuracy.


After analyzing the result, the 100% accuracy given by SVM could indicates that there might be an issue with the data or the way the model is evaluated as it should be quite rare to get 100% accuracy. Then, from each of the classification report, we could see that the Random Forest in CountVectorizer performs really well as the accuracy and classification metrics shows a pretty high number. So, the text representation and machine learning algorithm that performs the best on predicting is the CountVectorizer using Random Forest with 85.7% accuracy.
